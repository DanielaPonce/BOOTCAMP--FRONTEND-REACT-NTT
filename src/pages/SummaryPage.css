.summary-container {
	margin: 6rem 15rem;
	display: flex;
	flex-direction: column;
	justify-content: center;
	align-items: center;
	gap: 3rem;
}

@media only screen and (min-width: 740px) and (max-width: 1024px) {
	.summary-container {
		margin: 6rem 2rem;
	}
}
@media only screen and (max-width: 739px) {
	.summary-container {
		margin: 3rem 2rem;
	}
}
