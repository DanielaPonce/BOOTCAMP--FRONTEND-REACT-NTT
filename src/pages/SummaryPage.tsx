import CartSummary from '../components/Cart/CartSummary';
import CartTable from '../components/Cart/CartTable';
import Footer from '../components/Footer/Footer';
import Header from '../components/Header/Header';
import ShippingForm from '../components/ShippingForm/ShippingForm';

const SummaryPage = () => {
	return (
		<>
			<Header />
			<div className="summary-container">
				<CartTable />
				<CartSummary />
				<ShippingForm />
			</div>
			<Footer />
		</>
	);
};

export default SummaryPage;
