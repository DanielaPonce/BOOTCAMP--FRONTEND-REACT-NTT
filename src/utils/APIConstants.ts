export enum APIConstants {
	API_BASE_URL = 'https://dummyjson.com'
}
