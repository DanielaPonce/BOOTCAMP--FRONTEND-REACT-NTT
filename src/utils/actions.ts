export enum Actions {
	ADD_TO_CART_TYPE = 'ADD_TO_CART',
	REMOVE_FROM_CART_TYPE = 'REMOVE_FROM_CART',
	UPDATE_QUANTITY_TYPE = 'UPDATE_QUANTITY',
	CLEAR_CART_TYPE = 'CLEAR_CART'
}
