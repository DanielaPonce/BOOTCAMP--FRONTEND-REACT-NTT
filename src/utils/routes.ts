export enum RoutesConstants {
	home = '/',
	summary = 'summary'
}
