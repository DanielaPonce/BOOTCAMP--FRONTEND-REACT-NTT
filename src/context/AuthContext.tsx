import {
	createContext,
	FC,
	PropsWithChildren,
	useContext,
	useEffect,
	useState
} from 'react';
import { loginRequest } from '../api/authentication.api';
import { storageUtils } from '../utils/storageUtils';

interface AuthContextProps {
	isAuthenticated: boolean;
	user: string | null;
	isLoading: boolean;
	login: (username: string, password: string) => void;
	logout: () => void;
}

const AuthContext = createContext<AuthContextProps | undefined>(undefined);

export const AuthProvider: FC<PropsWithChildren> = ({ children }) => {
	const [isAuthenticated, setIsAuthenticated] = useState(false);
	const [user, setUser] = useState<string | null>(null);
	const [isLoading, setIsLoading] = useState(true);

	useEffect(() => {
		const storedUser = storageUtils.getUser();
		if (storedUser) {
			setIsAuthenticated(true);
			setUser(storedUser);
		}
		setIsLoading(false);
	}, []);

	const login = async (username: string, password: string) => {
		try {
			const loginData = await loginRequest(username, password);
			setIsAuthenticated(true);
			setUser(loginData.username);
			storageUtils.setUser(loginData.username);
		} catch (error) {
			setIsAuthenticated(false);
			setUser(null);
			console.log(error);
			throw new Error(
				error instanceof Error ? error.message : 'Login failed'
			);
		}
	};

	const logout = () => {
		setIsAuthenticated(false);
		setUser(null);
		storageUtils.clearUser();
	};

	return (
		<AuthContext.Provider
			value={{ isAuthenticated, user, isLoading, login, logout }}
		>
			{children}
		</AuthContext.Provider>
	);
};

export const useAuth = (): AuthContextProps => {
	const context = useContext(AuthContext);
	if (!context) {
		throw new Error('useAuth must be used within an AuthProvider');
	}
	return context;
};
