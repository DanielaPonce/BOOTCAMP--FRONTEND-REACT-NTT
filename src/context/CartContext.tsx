import {
	createContext,
	Dispatch,
	FC,
	ReactNode,
	useContext,
	useEffect,
	useReducer
} from 'react';
import { Product } from '../models/product.types';
import { Actions } from '../utils/actions';
import { LocalStorageConstants } from '../utils/localStorageConstants';

type CartAction =
	| { type: Actions.ADD_TO_CART_TYPE; payload: Product }
	| { type: Actions.REMOVE_FROM_CART_TYPE; payload: number }
	| {
			type: Actions.UPDATE_QUANTITY_TYPE;
			payload: { id: number; quantity: number };
	  }
	| { type: Actions.CLEAR_CART_TYPE };

export interface CartItemType extends Product {
	quantity: number;
}

interface CartState {
	cart: CartItemType[];
	total: number;
	itemCount: number;
}

const initialState: CartState = {
	cart: [],
	total: 0,
	itemCount: 0
};

const cartReducer = (state: CartState, action: CartAction): CartState => {
	switch (action.type) {
		case Actions.ADD_TO_CART_TYPE: {
			const existingItem = state.cart.find(
				(item) => item.id === action.payload.id
			);
			if (existingItem) {
				const updatedCart = state.cart.map((item) =>
					item.id === action.payload.id
						? { ...item, quantity: item.quantity + 1 }
						: item
				);
				const updatedTotal = updatedCart.reduce(
					(sum, item) => sum + item.price * item.quantity,
					0
				);
				const updatedItemCount = updatedCart.reduce(
					(sum, item) => sum + item.quantity,
					0
				);

				return {
					...state,
					cart: updatedCart,
					total: updatedTotal,
					itemCount: updatedItemCount
				};
			}

			const newCartItem: CartItemType = {
				...action.payload,
				quantity: 1
			};
			return {
				...state,
				cart: [...state.cart, newCartItem],
				total: state.total + action.payload.price,
				itemCount: state.itemCount + 1
			};
		}
		case Actions.REMOVE_FROM_CART_TYPE: {
			const filteredCart = state.cart.filter(
				(item) => item.id !== action.payload
			);
			// falta cubrir este caso
			const updatedTotal = filteredCart.reduce(
				(sum, item) => sum + item.price * item.quantity,
				0
			);
			// falta cubrir este caso
			const updatedItemCount = filteredCart.reduce(
				(sum, item) => sum + item.quantity,
				0
			);

			return {
				...state,
				cart: filteredCart,
				total: updatedTotal,
				itemCount: updatedItemCount
			};
		}
		case Actions.UPDATE_QUANTITY_TYPE: {
			const updatedCart = state.cart.map((item) =>
				item.id === action.payload.id
					? { ...item, quantity: action.payload.quantity }
					// falta cubrir este caso
					: item
			);
			const updatedTotal = updatedCart.reduce(
				(sum, item) => sum + item.price * item.quantity,
				0
			);
			const updatedItemCount = updatedCart.reduce(
				(sum, item) => sum + item.quantity,
				0
			);

			return {
				...state,
				cart: updatedCart,
				total: updatedTotal,
				itemCount: updatedItemCount
			};
		}
		case Actions.CLEAR_CART_TYPE:
			return initialState;
		// falta cubrir este caso
		default:
			throw new Error(`Acci√≥n desconocida: ${action}`);
	}
};

interface CartContextProps {
	state: CartState;
	dispatch: Dispatch<CartAction>;
}

const CartContext = createContext<CartContextProps | undefined>(undefined);

export const CartProvider: FC<{ children: ReactNode }> = ({ children }) => {
	const storedCart = localStorage.getItem(LocalStorageConstants.CartKey);
	const initialCartState: CartState = storedCart
		? JSON.parse(storedCart)
		: initialState;

	const [state, dispatch] = useReducer(cartReducer, initialCartState);

	return (
		<CartContext.Provider value={{ state, dispatch }}>
			{children}
		</CartContext.Provider>
	);
};

export const useCartContext = (): CartContextProps => {
	const context = useContext(CartContext);
	if (!context) {
		throw new Error('useCartContext debe usarse dentro de un CartProvider');
	}
	return context;
};
