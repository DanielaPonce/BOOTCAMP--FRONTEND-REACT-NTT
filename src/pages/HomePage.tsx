import { FC, useState } from 'react';
import Footer from '../components/Footer/Footer';
import Header from '../components/Header/Header';
import Paginator from '../components/ProductList/Paginator';
import ProductList from '../components/ProductList/ProductList';
import Spinner from '../components/Spinner/Spinner';
import { CartProvider } from '../context/CartContext';
import { useProducts } from '../hooks/useProducts';
import { Constants } from '../utils/constants';

const HomePage: FC = () => {
	const [search, setSearch] = useState('');
	const [category, setCategory] = useState(
		Constants.defaultSelectValue.valueOf()
	);

	const {
		products,
		isLoading,
		totalPages,
		currentPage,
		setCurrentPage,
		showPagination
	} = useProducts(category, search);

	return (
		<CartProvider>
			<Header onSearch={setSearch} onFilterCategory={setCategory} />
			{isLoading ? (
				<Spinner />
			) : (
				<div className="body-content">
					<ProductList products={products} />
					{showPagination ? (
						<Paginator
							currentPage={currentPage}
							totalPages={totalPages}
							onPageChange={setCurrentPage}
						/>
					) : (
						<></>
					)}
				</div>
			)}
			<Footer />
		</CartProvider>
	);
};

export default HomePage;
