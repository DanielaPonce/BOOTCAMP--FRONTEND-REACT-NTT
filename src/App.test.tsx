import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { MemoryRouter, Route, Routes } from 'react-router';
import { CartProvider } from './context/CartContext';
import HomePage from './pages/HomePage';
import SummaryPage from './pages/SummaryPage';
import { RoutesConstants } from './utils/routes';

jest.mock('./pages/HomePage', () => ({
	__esModule: true,
	default: jest.fn(() => {
		const { Link } = require('react-router');
		return (
			<div>
				HomePage
				<Link to={RoutesConstants.Summary}>Go to Summary</Link>
			</div>
		);
	})
}));

jest.mock('./pages/SummaryPage', () => ({
	__esModule: true,
	default: jest.fn(() => {
		const { Link } = require('react-router');
		return (
			<div>
				SummaryPage
				<Link to="/">Go to Home</Link>
			</div>
		);
	})
}));

describe('App Component', () => {
	const renderWithRouter = (initialEntries: string[]) => {
		render(
			<MemoryRouter initialEntries={initialEntries}>
				<CartProvider>
					<Routes>
						<Route path="/" element={<HomePage />} />
						<Route
							path={RoutesConstants.Summary}
							element={<SummaryPage />}
						/>
					</Routes>
				</CartProvider>
			</MemoryRouter>
		);
	};

	beforeEach(() => {
		jest.clearAllMocks();
	});

	test('renders HomePage by default', () => {
		renderWithRouter(['/']);
		expect(screen.getByText('HomePage')).toBeInTheDocument();
		expect(screen.queryByText('SummaryPage')).not.toBeInTheDocument();
	});

	test('renders SummaryPage when navigating to /summary', () => {
		renderWithRouter([RoutesConstants.Summary]);
		expect(screen.getByText('SummaryPage')).toBeInTheDocument();
		expect(screen.queryByText('HomePage')).not.toBeInTheDocument();
	});

	test('navigates between HomePage and SummaryPage using links', async () => {
		const user = userEvent.setup();

		renderWithRouter(['/']);
		expect(screen.getByText('HomePage')).toBeInTheDocument();

		await user.click(screen.getByText('Go to Summary'));
		expect(screen.getByText('SummaryPage')).toBeInTheDocument();
		expect(screen.queryByText('HomePage')).not.toBeInTheDocument();

		await user.click(screen.getByText('Go to Home'));
		expect(screen.getByText('HomePage')).toBeInTheDocument();
		expect(screen.queryByText('SummaryPage')).not.toBeInTheDocument();
	});

	test('matches snapshot for HomePage', () => {
		const { asFragment } = render(
			<MemoryRouter initialEntries={['/']}>
				<CartProvider>
					<HomePage />
				</CartProvider>
			</MemoryRouter>
		);
		expect(asFragment()).toMatchSnapshot();
	});

	test('matches snapshot for SummaryPage', () => {
		const { asFragment } = render(
			<MemoryRouter initialEntries={[RoutesConstants.Summary]}>
				<CartProvider>
					<SummaryPage />
				</CartProvider>
			</MemoryRouter>
		);
		expect(asFragment()).toMatchSnapshot();
	});
});
