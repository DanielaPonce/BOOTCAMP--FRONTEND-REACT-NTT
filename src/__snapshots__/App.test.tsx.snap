// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component matches snapshot for HomePage 1`] = `
<DocumentFragment>
  <div>
    HomePage
    <a
      data-discover="true"
      href="/summary"
    >
      Go to Summary
    </a>
  </div>
</DocumentFragment>
`;

exports[`App Component matches snapshot for SummaryPage 1`] = `
<DocumentFragment>
  <div>
    SummaryPage
    <a
      data-discover="true"
      href="/"
    >
      Go to Home
    </a>
  </div>
</DocumentFragment>
`;
